smithy-model/src/main/java/software/amazon/smithy/model/traits/HttpChecksumTrait.java

```diff
@@ -0,0 +1,284 @@
+                .responseProperty(responseProperty);
+    }
+
+    /**
+     * Gets request property defined within the HttpChecksum trait.
+     *
+     * @return checksum properties for request.
+     */
+    public HttpChecksumProperties getRequestProperty() {
+        return requestProperty;
+    }
+
+    /**
+     * Gets response property defined within the HttpChecksum trait.
+     *
+     * @return checksum properties for response.
+     */
+    public HttpChecksumProperties getResponseProperty() {
+        return responseProperty;
+    }
+
+    @Override
+    protected Node createNode() {
+        ObjectNode.Builder builder = ObjectNode.objectNodeBuilder();
+        builder.sourceLocation(getSourceLocation());
+
+        if (requestProperty != null) {
+            builder.withMember(REQUEST_PROPERTY, requestProperty.toNode());
+        }
+        if (responseProperty != null) {
+            builder.withMember(RESPONSE_PROPERTY, responseProperty.toNode());
+        }
+        return builder.build();
+    }
+
+
+    public static final class Builder extends AbstractTraitBuilder<HttpChecksumTrait, Builder> {
+        private HttpChecksumProperties requestProperty;
+        private HttpChecksumProperties responseProperty;
+
+        private Builder() {
+        }
+
+        @Override
+        public HttpChecksumTrait build() {
+            return new HttpChecksumTrait(this);
+        }
+
+        public Builder requestProperty(HttpChecksumProperties requestProperty) {
+            this.requestProperty = requestProperty;
+            return this;
+        }
+
+        public Builder responseProperty(HttpChecksumProperties responseProperty) {
+            this.responseProperty = responseProperty;
+            return this;
+        }
+    }
+
+    public static final class Provider implements TraitService {
+        @Override
+        public ShapeId getShapeId() {
+            return ID;
+        }
+
+        public Trait createTrait(ShapeId target, Node value) {
+            ObjectNode node = value.expectObjectNode();
+            Builder builder = builder().sourceLocation(value);
+
+            Optional<ObjectNode> requestNode = node.getObjectMember(REQUEST_PROPERTY);
+            if (requestNode.isPresent()) {
+                builder.requestProperty = HttpChecksumProperties.fromNode(requestNode.get());
+            }
+
+            Optional<ObjectNode> responseNode = node.getObjectMember(RESPONSE_PROPERTY);
+            if (responseNode.isPresent()) {
+                builder.responseProperty = HttpChecksumProperties.fromNode(responseNode.get());
+            }
+
+            return builder.build();
+        }
+    }
+
+
+    /**
+     * Defines checksum properties for data checksum validation.
+     * These properties are used by the members defined for HTTPChecksum trait.
+     */
+    public static final class HttpChecksumProperties implements ToNode, ToSmithyBuilder<HttpChecksumProperties> {
```

Comment:
Keeping `HttpChecksumProperties` as an inner class of `HttpChecksumTrait` as `software.amazon.smithy.model.traits` doesn't seem to be the correct namespace for this class (mostly as this class isn't a trait in itself). 

Please let me know if I should move this, or it is okay to keep it as an inner class.

Reply:
I'd say break this out as there is precedence for having a supporting class defined in the `*.traits` package. For example, [the `EnumDefinition` class](https://github.com/awslabs/smithy/blob/f737025ce0d389409a728b439ab368f3f9b47d23/smithy-model/src/main/java/software/amazon/smithy/model/traits/EnumDefinition.java) is defined in the `software.amazon.smithy.model.traits` package even though it's not a trait itself.